crates/sui-adapter/src/programmable_transactions/execution.rs

```diff
@@ -460,17 +463,184 @@ fn execute_move_publish<E: fmt::Debug, S: StorageView<E>, Mode: ExecutionMode>(
+        !module_bytes.is_empty(),
+        "empty package is checked in transaction input checker"
+    );
+
+    let upgrade_ticket: UpgradeTicket = {
+        let mut ticket_bytes = Vec::new();
+        let ticket_val: Value =
+            context.by_value_arg(CommandKind::Upgrade, 0, upgrade_ticket_arg)?;
+        let ticket_type = context
+            .session
+            .load_type(&TypeTag::Struct(Box::new(UpgradeTicket::type_())))
+            .map_err(|e| context.convert_vm_error(e))?;
+        check_param_type::<_, _, Mode>(context, 0, &ticket_val, &ticket_type)?;
+        ticket_val.write_bcs_bytes(&mut ticket_bytes);
+        bcs::from_bytes(&ticket_bytes).map_err(|_| {
+            ExecutionError::from_kind(ExecutionErrorKind::CommandArgumentError {
+                arg_idx: 0,
+                kind: CommandArgumentError::InvalidBCSBytes,
+            })
+        })?
+    };
+
+    // Check digest.
+    let computed_digest =
+        MovePackage::compute_digest_for_modules_and_deps(&module_bytes, &dep_ids).to_vec();
+    if computed_digest != upgrade_ticket.digest {
+        return Err(ExecutionError::from_kind(
+            ExecutionErrorKind::PackageUpgradeError {
+                upgrade_error: PackageUpgradeError::DigestDoesNotMatch {
+                    digest: computed_digest,
+                },
+            },
+        ));
+    }
+
+    // Run the move + sui verifier on the modules and publish them into the cache.
+    // This will give use the new package ID as well that we will need when creating
+    // the linkage tables when upgrading the package.
+    let (upgraded_package_modules, new_package_id) =
+        publish_and_verify_modules::<_, _, Mode>(context, module_bytes)?;
+
+    // Check that this package ID points to a package and get the package we're upgrading.
+    let current_package = fetch_package(context, &upgrade_ticket.package.bytes)?;
+
+    // Full backwards compatibility except that we allow friend function signatures to change.
+    let compatiblity_checker = {
+        let check_struct_and_pub_function_linking = true;
+        let check_struct_layout = true;
+        let check_friend_linking = false;
+        Compatibility::new(
+            check_struct_and_pub_function_linking,
+            check_struct_layout,
+            check_friend_linking,
+        )
+    };
+    check_compatibility(
+        &current_package,
+        &upgraded_package_modules,
+        compatiblity_checker,
+        upgrade_ticket.policy,
+    )?;
+
+    // Deduplicate any dependency ids.
+    dep_ids.dedup();
+
+    // Read the package dependencies.
+    let dependency_packages = dep_ids
+        .iter()
+        .map(|id| fetch_package(context, id))
+        .collect::<Result<Vec<_>, _>>()?;
+
+    let upgraded_object_id = context.upgrade_package(
+        &current_package,
+        new_package_id,
+        upgraded_package_modules,
+        &dependency_packages,
+    )?;
+
+    Ok(vec![Value::Object(ObjectValue::new(
```

Comment:
woah, I didn't think `UpgradeReceipt` was an object?

Reply:
This can only be used for objects, which I don't think `UpgradeReceipt` or `UpgradeTicket` are? You have to use `Value::Raw` instead

Reply:
Good catch! Totally right this isn't an object -- I'll update